src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
>> ASP.NET in particular does not put this constraint

> Yes it does. Conforming containers, which is all we support, have well defined constructor selection, which is what we should follow.

Ah, I guess I now understand where the misconception is coming from. IServiceProvider will pick a specific constructor, but only if the registration just identified the type. But of course you can also even provide a factory method, so you don't even need an accessible constructor at all (not having a constructor is not supported in the default case, which uses ActivatorUtilities).

Also, the IControllerActivator extensibility point exists so you don't have to depend neither on ActivatorUtilities nor IServiceProvider at all. They even have an additional extensibility point so that if you are using the DefaultControllerActivator, you could use a different ITypeActivatorCache so that you can customize object creation (here is where the ActivatorUtilities is used by default actually).

I'm out, but if you can't find examples, on Wednesday I will be able to share some small samples of how to use different constructors or containers in ASP.NET. I'm 100% it is the case that ASP.NET doesn't put that constraint on which constructor to use from a controller. They have the same default as our default, but you can always change that.
See this article for some info on this extensibility point, as well as the ServiceBasedControllerActivator that ships in the box: https://andrewlock.net/controller-activation-and-dependency-injection-in-asp-net-core-mvc/

> Just a convenient place to cache grain type specific data. This is an internal implementation detail. If we find a better place we can move it.

Well, as I suggested, you can easily put it in a static field of `CachedReadConcurrentDictionary<Type, ParameterInfo[]>` on first use of that utility, and avoid adding the constraint and the coupling.